### ISSUE TITLE
Issues with using junit

### ISSUE THREAD

**Issue by kyonifer** [cbeust/kobalt#251]
I was playing around with kobalt to see if I could use it for my kotlin lib (https://github.com/kyonifer/golem/tree/kobalt). Using the intellij kobalt plugin works fine for running all the tests, as does "gradle test" on the version in master. However, when I try to run my junit tests from the command line with "kobalt test" it fails with (snipped):

```
...
java.lang.Exception: No runnable methods
	at org.junit.runners.BlockJUnit4ClassRunner.validateInstanceMethods(BlockJUnit4ClassRunner.java:191)
	at org.junit.runners.BlockJUnit4ClassRunner.collectInitializationErrors(BlockJUnit4ClassRunner.java:128)
	at org.junit.runners.ParentRunner.validate(ParentRunner.java:416)
	at org.junit.runners.ParentRunner.<init>(ParentRunner.java:84)
	at org.junit.runners.BlockJUnit4ClassRunner.<init>(BlockJUnit4ClassRunner.java:65)
	at org.junit.internal.builders.JUnit4Builder.runnerForClass(JUnit4Builder.java:10)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at org.junit.runner.Computer.getRunner(Computer.java:40)
	at org.junit.runner.Computer$1.runnerForClass(Computer.java:31)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:101)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:87)
	at org.junit.runners.Suite.<init>(Suite.java:81)
	at org.junit.runner.Computer.getSuite(Computer.java:28)
	at org.junit.runner.Request.classes(Request.java:75)
	at org.junit.runner.JUnitCommandLineParseResult.createRequest(JUnitCommandLineParseResult.java:118)
	at org.junit.runner.JUnitCore.runMain(JUnitCore.java:77)
	at org.junit.runner.JUnitCore.main(JUnitCore.java:36)
600) initializationError(golem.util.test.UtilTests$assertMatrixEquals$1)
java.lang.Exception: The class golem.util.test.UtilTests$assertMatrixEquals$1 is not public.
	at org.junit.validator.PublicClassValidator.validateTestClass(PublicClassValidator.java:29)
	at org.junit.runners.ParentRunner.applyValidators(ParentRunner.java:134)
	at org.junit.runners.ParentRunner.collectInitializationErrors(ParentRunner.java:128)
	at org.junit.runners.BlockJUnit4ClassRunner.collectInitializationErrors(BlockJUnit4ClassRunner.java:124)
	at org.junit.runners.ParentRunner.validate(ParentRunner.java:416)
	at org.junit.runners.ParentRunner.<init>(ParentRunner.java:84)
	at org.junit.runners.BlockJUnit4ClassRunner.<init>(BlockJUnit4ClassRunner.java:65)
	at org.junit.internal.builders.JUnit4Builder.runnerForClass(JUnit4Builder.java:10)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at org.junit.runner.Computer.getRunner(Computer.java:40)
	at org.junit.runner.Computer$1.runnerForClass(Computer.java:31)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:101)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:87)
	at org.junit.runners.Suite.<init>(Suite.java:81)
	at org.junit.runner.Computer.getSuite(Computer.java:28)
	at org.junit.runner.Request.classes(Request.java:75)
	at org.junit.runner.JUnitCommandLineParseResult.createRequest(JUnitCommandLineParseResult.java:118)
	at org.junit.runner.JUnitCore.runMain(JUnitCore.java:77)
	at org.junit.runner.JUnitCore.main(JUnitCore.java:36)
601) initializationError(golem.util.test.UtilTests$assertMatrixEquals$1)
java.lang.Exception: Test class should have exactly one public constructor
	at org.junit.runners.BlockJUnit4ClassRunner.validateOnlyOneConstructor(BlockJUnit4ClassRunner.java:158)
	at org.junit.runners.BlockJUnit4ClassRunner.validateConstructor(BlockJUnit4ClassRunner.java:147)
	at org.junit.runners.BlockJUnit4ClassRunner.collectInitializationErrors(BlockJUnit4ClassRunner.java:127)
	at org.junit.runners.ParentRunner.validate(ParentRunner.java:416)
	at org.junit.runners.ParentRunner.<init>(ParentRunner.java:84)
	at org.junit.runners.BlockJUnit4ClassRunner.<init>(BlockJUnit4ClassRunner.java:65)
	at org.junit.internal.builders.JUnit4Builder.runnerForClass(JUnit4Builder.java:10)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at org.junit.runner.Computer.getRunner(Computer.java:40)
	at org.junit.runner.Computer$1.runnerForClass(Computer.java:31)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:101)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:87)
	at org.junit.runners.Suite.<init>(Suite.java:81)
	at org.junit.runner.Computer.getSuite(Computer.java:28)
	at org.junit.runner.Request.classes(Request.java:75)
	at org.junit.runner.JUnitCommandLineParseResult.createRequest(JUnitCommandLineParseResult.java:118)
	at org.junit.runner.JUnitCore.runMain(JUnitCore.java:77)
	at org.junit.runner.JUnitCore.main(JUnitCore.java:36)
602) initializationError(golem.util.test.UtilTests$assertMatrixEquals$1)
java.lang.Exception: No runnable methods
	at org.junit.runners.BlockJUnit4ClassRunner.validateInstanceMethods(BlockJUnit4ClassRunner.java:191)
	at org.junit.runners.BlockJUnit4ClassRunner.collectInitializationErrors(BlockJUnit4ClassRunner.java:128)
	at org.junit.runners.ParentRunner.validate(ParentRunner.java:416)
	at org.junit.runners.ParentRunner.<init>(ParentRunner.java:84)
	at org.junit.runners.BlockJUnit4ClassRunner.<init>(BlockJUnit4ClassRunner.java:65)
	at org.junit.internal.builders.JUnit4Builder.runnerForClass(JUnit4Builder.java:10)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at org.junit.runner.Computer.getRunner(Computer.java:40)
	at org.junit.runner.Computer$1.runnerForClass(Computer.java:31)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:101)
	at org.junit.runners.model.RunnerBuilder.runners(RunnerBuilder.java:87)
	at org.junit.runners.Suite.<init>(Suite.java:81)
	at org.junit.runner.Computer.getSuite(Computer.java:28)
	at org.junit.runner.Request.classes(Request.java:75)
	at org.junit.runner.JUnitCommandLineParseResult.createRequest(JUnitCommandLineParseResult.java:118)
	at org.junit.runner.JUnitCore.runMain(JUnitCore.java:77)
	at org.junit.runner.JUnitCore.main(JUnitCore.java:36)

FAILURES!!!
Tests run: 689,  Failures: 602
```

Build file im trying out with is https://github.com/kyonifer/golem/blob/kobalt/kobalt/src/Build.kt . Any ideas on what could be up?

**Comment by cbeust** [comment]
I'm able to reproduce the problem. Kobalt is trying to run classes that are not test classes, some of which are downright weird:

```
golem.BuildersTests$testBuilder$1$expected$1
```

As far as I can tell, `expected` is a `var` in this class and yet, it seems to have generated a class.

**Comment by cbeust** [comment]
Do you know how many tests and which tests Gradle runs? I'm not getting any information with `./gradlew test -i`:

```
:test UP-TO-DATE
:test (Thread[Daemon worker Thread 2,5,main]) completed. Took 0.019 secs.

BUILD SUCCESSFUL

Total time: 0.838 secs
Stopped 0 compiler daemon(s).
Received result Success[value=null] from daemon DaemonInfo{pid=59944, address=[6e0d6259-0a5b-411d-972c-e0188abb6a61 port:65509, addresses:[/0:0:0:0:0:0:0:1, /127.0.0.1]], idle=true, context=DefaultDaemonContext[uid=ec7c808d-584b-455a-99f2-d90b468ddb34,javaHome=/Library/Java/JavaVirtualMachines/jdk1.8.0_73.jdk/Contents/Home,daemonRegistryDir=/Users/beust/.gradle/daemon,pid=59944,idleTimeout=10800000,daemonOpts=-XX:MaxPermSize=256m,-XX:+HeapDumpOnOutOfMemoryError,-Xmx1024m,-Dfile.encoding=UTF-8,-Duser.country=US,-Duser.language=en,-Duser.variant]} (build should be done).
```

**Comment by cbeust** [comment]
I think the bottom line is I need to be more precise in what classes I select to pass to JUnit or it gets easily upset. I'm thinking I'll limit this to classes that contain a JUnit annotation (any annotation).

**Comment by kyonifer** [comment]
Should be 87 tests (thats how many gradle picks up on master).

**Comment by cbeust** [comment]
Ok, I'll get back to you.

**Comment by kyonifer** [comment]
Sounds great, thanks for looking into it.

**Comment by cbeust** [comment]
Should be fixed in the latest:

```
$ ./kobaltw update
$ ./kobaltw test
            _  __          _               _   _
           | |/ /   ___   | |__     __ _  | | | |_
           | ' /   / _ \  | '_ \   / _` | | | | __|
           | . \  | (_) | | |_) | | (_| | | | | |_
           |_|\_\  \___/  |_.__/   \__,_| |_|  \__|  0.833

          ╔════════════════╗
          ║ Building golem ║
          ╚════════════════╝
───── golem:compile
  Kotlin 1.0.3 compiling 33 directories
───── golem:compileTest
  Kotlin 1.0.3 compiling 41 directories
───── golem:test
JUnit version 4.12
............Jun 30, 2016 4:13:59 AM com.github.fommil.jni.JniLoader liberalLoad
INFO: successfully loaded /var/folders/77/kjr_lq4x5tj5ymxdfvxs6c7c002p8z/T/jniloader5666141313130455788netlib-native_system-osx-x86_64.jnilib
Jun 30, 2016 4:13:59 AM com.github.fommil.jni.JniLoader load
INFO: already loaded netlib-native_system-osx-x86_64.jnilib
...........................................................................
Time: 0.552

OK (87 tests)

All tests passed
BUILD SUCCESSFUL (12 seconds)
```

Note that the tests try to download dependencies that I don't have and which seem to be defined in your local maven only:

```
Downloading: https://jcenter.bintray.com/de/erichseifert/vectorgraphics2d/VectorGraphics2D/0.9.2/VectorGraphics2D-0.9.2.pom
```

Let me know if this is fixed for you.

**Comment by kyonifer** [comment]
Works great now, closing this.

Re: the dependency, if I remove it from the ~/.kobalt cache, somehow kobalt still finds it and downloads it for me again:

```
./kobaltw test                                                            
            _  __          _               _   _   
           | |/ /   ___   | |__     __ _  | | | |_ 
           | ' /   / _ \  | '_ \   / _` | | | | __|
           | . \  | (_) | | |_) | | (_| | | | | |_ 
           |_|\_\  \___/  |_.__/   \__,_| |_|  \__|  0.833

          ╔════════════════╗
          ║ Building golem ║
          ╚════════════════╝
───── golem:compile
───── golem:compileTest
Downloaded artifact de.erichseifert.vectorgraphics2d:VectorGraphics2D:pom:0.9.2 from erichseifert.de (http://mvn.erichseifert.de/maven2, default, releases+snapshots)
Downloaded artifact de.erichseifert.vectorgraphics2d:VectorGraphics2D:jar:0.9.2 from erichseifert.de (http://mvn.erichseifert.de/maven2, default, releases+snapshots)
  Kotlin 1.0.3 compiling 41 directories
───── golem:test
JUnit version 4.12
.......Jul 01, 2016 12:25:09 AM com.github.fommil.jni.JniLoader liberalLoad
INFO: successfully loaded /tmp/jniloader4150554050175584413netlib-native_system-linux-x86_64.so
.......Jul 01, 2016 12:25:09 AM com.github.fommil.jni.JniLoader load
INFO: already loaded netlib-native_system-linux-x86_64.so
.........................................................................
Time: 0.527

OK (87 tests)

All tests passed
BUILD SUCCESSFUL (33 seconds)
```

I thought that was curious since I hadn't told it that repo name yet. Not sure whats going on there. Is kobalt (despite the message) actually polling a local gradle/ivy/mvn cache I haven't cleared? Grabbing a repo name from a pom? (Is there an easy way to attach an intellij debugger to kobaltw to see what its doing?) Its actually an optional dependency in mvn and I don't need it for golem, so in gradle I was using:

```
exclude module: 'VectorGraphics2D'
```

**Comment by kyonifer** [comment]
If I depend on the latest xchart 3.1.0, kobalt pulls in the dependency properly from the newer version of vectorgraphics2d that's hosted on jcenter:

```
Downloaded artifact org.knowm.xchart:xchart:pom:3.1.0 from maven (https://jcenter.bintray.com/, default, releases+snapshots)
Downloaded artifact org.knowm.xchart:xchart-parent:pom:3.1.0 from maven (https://jcenter.bintray.com/, default, releases+snapshots)
Downloaded artifact de.erichseifert.vectorgraphics2d:VectorGraphics2D:pom:0.11 from maven (https://jcenter.bintray.com/, default, releases+snapshots)
Downloaded artifact de.erichseifert.vectorgraphics2d:VectorGraphics2D:jar:0.11 from maven (https://jcenter.bintray.com/, default, releases+snapshots)
Downloaded artifact org.knowm.xchart:xchart:jar:3.1.0 from maven (https://jcenter.bintray.com/, default, releases+snapshots)
```

So probably the path forward for me is to fix my code to work with the latest xchart and avoid the weird old dependency. I'm still curious as to how kobalt knew where to find it for me and not for you though.

**Comment by cbeust** [comment]
Odd, I couldn't find it on JCenter, I must have missed it. You can also exclude with Kobalt, if that's how you want to fix that.

Glad to hear everything is working, thanks for the report!

**Comment by kyonifer** [comment]
Looks like the latest breaks the tests again (it tries to run 689 tests instead of 87). Doing a binary search it appears again on 0.835, so probably a regression due to https://github.com/cbeust/kobalt/commit/4d23c1236bba30f2d745f9c51638bc2d85653f3e

To keep it apples-to-apples im testing on the original golem commit I opened the issue under, https://github.com/kyonifer/golem/commit/11cff9b5af8fd8c260880d726145a77b978e54c6. Latest golem-kobalt branch HEAD is doing some multiproject stuff and it also appears there.

**Comment by cbeust** [comment]
Thanks, will investigate.

-- 
Cédric

**Comment by cbeust** [comment]
Fixed in 0.853.
